satpy/readers/modis_l3.py

```diff
@@ -0,0 +1,162 @@
+
+        # Initialise number of rows and columns
+        self.nrows = self.metadata["GridStructure"]["GRID_1"]["YDim"]
+        self.ncols = self.metadata["GridStructure"]["GRID_1"]["XDim"]
+
+        # Get the grid name and other projection info
+        self.area_extent = self._sort_grid()
+
+
+    def _sort_grid(self):
+        """Get the grid properties."""
+        # First, get the grid resolution
+        self._get_res()
+
+        # Now compute the data extent
+        upperleft = self.metadata["GridStructure"]["GRID_1"]["UpperLeftPointMtrs"]
+        lowerright = self.metadata["GridStructure"]["GRID_1"]["LowerRightMtrs"]
+
+        # For some reason, a few of the CMG products multiply their
+        # decimal degree extents by one million. This fixes it.
+        if lowerright[0] > 1e6 or upperleft[0] > 1e6:
+            upperleft = tuple(val / 1e6 for val in upperleft)
+            lowerright = tuple(val / 1e6 for val in lowerright)
+
+        return upperleft[0], lowerright[1], lowerright[0], upperleft[1]
+
+
+    def _get_res(self):
+        """Compute the resolution from the file metadata."""
+        gridname = self.metadata["GridStructure"]["GRID_1"]["GridName"]
+        if "CMG" not in gridname:
+            raise ValueError("Only CMG grids are supported")
+
+        # Get the grid resolution from the grid name
+        pos = gridname.rfind("_") + 1
+        pos2 = gridname.rfind("Deg")
+
+        # Some products don't have resolution listed.
+        if pos < 0 or pos2 < 0:
+            self.resolution = 360. / self.ncols
+        else:
+            self.resolution = float(gridname[pos:pos2])
+
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+
+        # Initialise set of variable names to carry through code
+        handled_var_names = set()
+
+        yield from super().available_datasets(configured_datasets)
+
+        ds_dict = self.sd.datasets()
+
+        for is_avail, ds_info in (configured_datasets or []):
+            file_key = ds_info.get("file_key", ds_info["name"])
+            # we must add all variables here even if another file handler has
+            # claimed the variable. It could be another instance of this file
+            # type and we don't want to add that variable dynamically if the
+            # other file handler defined it by the YAML definition.
+            handled_var_names.add(file_key)
+            if is_avail is not None:
+                # some other file handler said it has this dataset
+                # we don't know any more information than the previous
+                # file handler so let's yield early
+                yield is_avail, ds_info
+                continue
+            if self.file_type_matches(ds_info["file_type"]) is None:
+                # this is not the file type for this dataset
+                yield None, ds_info
```

Comment:
I think this is a bug in the `viirs_edr.py` reader, but I think this should be followed by a `continue`. The reason being that if we aren't the file for this data set (file type doesn't match) then we shouldn't be claiming anything about it (the below `yield file_key in ds_dict.keys(), ds_info` line is doing this).

Unless you disagree, could you put a `continue` below this yield and fix the `viirs_edr.py` reader? To trigger this bug you'd have to process a file type that did not match one of the YAML-defined datasets, but even then the end result for the user would be the same (the variable isn't available). :thinking: I guess if you loaded multiple file types and only one of them matched the YAML-defined variable and this...no I think the behavior would be the same.

Reply:
I just asked about this on slack as my tests brought up this problem (the same DS was being returned twice, one with `None` and one with `False`). Will add in the `continue`